File: mobile-app/lib/services/activity_log_actions.dart
/// Canonical activity log actions used for auditing and report analytics.
class ActivityLogActions {
  ActivityLogActions._();

  static const login = 'Login';
  static const logout = 'Logout';
  static const bookAppointment = 'Book Appointment';
  static const updateAppointment = 'Update Appointment';
  static const updateAppointmentStatus = 'Update Appointment Status';
  static const acceptAppointment = 'Accept Appointment';
  static const completeAppointment = 'Complete Appointment';
  static const updatePatient = 'Update Patient Record';
  static const registerAccount = 'Register Account';
  static const sendMessage = 'Send Message';
  static const markMedication = 'Mark Medication Taken';
  static const emergencyAlert = 'Emergency Alert Submitted';
  static const voiceCall = 'Voice Call Made';
  static const healthRecordViewed = 'Health Record Viewed';
  static const failedGps = 'Failed GPS Retrieval';
  static const cameraDenied = 'Camera Permission Denied';
  static const voiceRecognitionFailure = 'Voice Recognition Failure';
  static const networkTimeout = 'Network Timeout';
  static const failedLogin = 'Failed Login';
  static const unauthorizedAccess = 'Unauthorized Access Attempt';
  static const passwordChange = 'Password Change';
  static const accountLockout = 'Account Lockout';
  static const respondEmergency = 'Respond to Emergency';
  static const assignCaregiver = 'Assign Caregiver';
  static const resolveEmergency = 'Resolve Emergency Alert';
}

File: admin-web/js/activity-log-actions.js
/** Canonical activity log actions used for auditing and report analytics. */
export const LOG_ACTIONS = {
  LOGIN: "Login",
  LOGOUT: "Logout",
  BOOK_APPOINTMENT: "Book Appointment",
  UPDATE_APPOINTMENT: "Update Appointment",
  UPDATE_APPOINTMENT_STATUS: "Update Appointment Status",
  ACCEPT_APPOINTMENT: "Accept Appointment",
  COMPLETE_APPOINTMENT: "Complete Appointment",
  UPDATE_PATIENT: "Update Patient Record",
  CREATE_PATIENT: "Create Patient Account",
  CREATE_STAFF: "Create Staff Account",
  UPDATE_STAFF: "Update Staff Account",
  REGISTER_ACCOUNT: "Register Account",
  SEND_MESSAGE: "Send Message",
  MARK_MEDICATION: "Mark Medication Taken",
  EMERGENCY_ALERT: "Emergency Alert Submitted",
  VOICE_CALL: "Voice Call Made",
  HEALTH_RECORD_VIEWED: "Health Record Viewed",
  FAILED_GPS: "Failed GPS Retrieval",
  CAMERA_DENIED: "Camera Permission Denied",
  VOICE_RECOGNITION_FAILURE: "Voice Recognition Failure",
  NETWORK_TIMEOUT: "Network Timeout",
  FAILED_LOGIN: "Failed Login",
  UNAUTHORIZED_ACCESS: "Unauthorized Access Attempt",
  PASSWORD_CHANGE: "Password Change",
  ACCOUNT_LOCKOUT: "Account Lockout",
  RESPOND_EMERGENCY: "Respond to Emergency",
  ASSIGN_CAREGIVER: "Assign Caregiver",
  RESOLVE_EMERGENCY: "Resolve Emergency Alert",
};

export function computeReportAnalytics(logs) {
  const countAction = (action) =>
    logs.filter((log) => log.action === action).length;

  return {
    userActivity: {
      totalLogins: countAction(LOG_ACTIONS.LOGIN),
      totalAppointmentsBooked: countAction(LOG_ACTIONS.BOOK_APPOINTMENT),
      totalEmergencyAlerts: countAction(LOG_ACTIONS.EMERGENCY_ALERT),
      totalVoiceCalls: countAction(LOG_ACTIONS.VOICE_CALL),
      totalHealthRecordsViewed: countAction(LOG_ACTIONS.HEALTH_RECORD_VIEWED),
    },
    warning: {
      failedGps: countAction(LOG_ACTIONS.FAILED_GPS),
      cameraDenied: countAction(LOG_ACTIONS.CAMERA_DENIED),
      voiceRecognitionFailure: countAction(LOG_ACTIONS.VOICE_RECOGNITION_FAILURE),
      networkTimeout: countAction(LOG_ACTIONS.NETWORK_TIMEOUT),
    },
    security: {
      failedLogin: countAction(LOG_ACTIONS.FAILED_LOGIN),
      unauthorizedAccess: countAction(LOG_ACTIONS.UNAUTHORIZED_ACCESS),
      passwordChanges: countAction(LOG_ACTIONS.PASSWORD_CHANGE),
      accountLockouts: countAction(LOG_ACTIONS.ACCOUNT_LOCKOUT),
    },
  };
}
